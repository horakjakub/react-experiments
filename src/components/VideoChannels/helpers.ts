const thumbnailUrl1 =
  'https://yt3.ggpht.com/-_p81gJCwoJA/AAAAAAAAAAI/AAAAAAAAAAA/1j_w8z5jhxs/s800-c-k-no-mo-rj-c0xffffff/photo.jpg';
const thumbnailUrl2 =
  'https://yt3.ggpht.com/-HMwHc6Xhc6A/AAAAAAAAAAI/AAAAAAAAAAA/91QtuNLyRz4/s800-c-k-no-mo-rj-c0xffffff/photo.jpg';
const thumbnailUrl3 =
  'https://yt3.ggpht.com/-eddbsyN7uAw/AAAAAAAAAAI/AAAAAAAAAAA/pLdaSejwsnk/s800-c-k-no-mo-rj-c0xffffff/photo.jpg';
const thumbnailUrl4 =
  'https://yt3.ggpht.com/-pCVFPJc8vX0/AAAAAAAAAAI/AAAAAAAAAAA/d7XQ8FOSt20/s800-c-k-no-mo-rj-c0xffffff/photo.jpg';

function getThumbnail(no: number) {
  switch (no) {
    case 1:
      return thumbnailUrl1;
    case 2:
      return thumbnailUrl2;
    case 3:
      return thumbnailUrl3;
    default:
      return thumbnailUrl4;
  }
}

let count = 0;
export function generateMockVideoChannelData(id: string) {
  if (count === 4) {
    count = 0;
  }
  count++;
  return {
    id: `cat ${id}`,
    title: `Cool title ${id}`,
    description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
      Nunc eget mass nunc. Nulla tellus massa, malesuada vulputate velit
      nec, mollis vulputate tellus. Integer pretium vulputate purus, et
      vulputate odio egestas non. Morbi ultrices lectus eu blandit euismod.
       Nunc eget massa nunc. Nulla tellus massa, malesuada vulputate velit
      nec, mollis vulputate tellus. Integer pretium vulputate purus, et
      vulputate odio egestas non. Morbi ultrices lectus eu blandit euismod.  Donec orci massa, feugiat nec maximus non, tristique at velit.
      Phasellus hendrerit odio ut lobortis faucibus. `,
    thumbnailUrl: getThumbnail(count),
  };
}
