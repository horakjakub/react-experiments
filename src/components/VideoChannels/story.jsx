import React from 'react';
import GlobalStyle from 'global-styles';

export default {
  title: 'VideoChannels',
};

export const DefaultVideoChannels = () => (
  <>
    <GlobalStyle />
  </>
);

export const VideoChannelsWithLoader = () => (
  <>
    <GlobalStyle />
  </>
);

export const VideoChannelsWithNoMoreResults = () => (
  <>
    <GlobalStyle />
  </>
);


