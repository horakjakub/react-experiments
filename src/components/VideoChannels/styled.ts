import styled from 'styled-components';

export const VideoChannelsBox = styled.div``;

export const VideoChannelsGrid = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(36em, 1fr));
  gap: 2em;
  margin: 1em;
`;

export const ResultsPlaceholder = styled.div`
  width: 36em;
  height: 44em;
  margin: 1.5em 1em 1em 1em;
  background: rgba(255, 255, 255, 0.8);
  display: flex;
  justify-content: center;
  align-items: center;
  p {
    font-size: 2.4em;
    font-family: 'Lato', sans-serif;
    color: pink;
    padding: 1em;
    text-align: center;
  }
`;
