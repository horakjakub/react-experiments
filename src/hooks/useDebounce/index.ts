import {useState, useEffect} from 'react';
import {debounce} from 'lodash';

export default useDebounce;

function useDebounce(phrase: string | null): {debouncedPhrase: string | null} {
  const [debouncedPhrase, setDebouncedPhrase] = useState<string | null>(null);

  const setDebounced = debounce(txt => setDebouncedPhrase(txt), 300);

  useEffect(() => {
    setDebounced(phrase);
  }, [phrase, setDebounced]);

  return {debouncedPhrase};
}
