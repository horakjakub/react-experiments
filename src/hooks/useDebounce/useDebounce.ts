import React, { useState, useEffect } from 'react';
import { debounce } from 'lodash';

export default useDebounce;

function useDebounce(
  phrase: string | null
): { debouncedPhrase: string | null } {
  const [debouncedPhrase, setDebouncedPhrase] = useState<string | null>(null);

  useEffect(
    debounce(() => {
      setDebouncedPhrase(phrase);
    }, 300)
  );

  return { debouncedPhrase };
}
