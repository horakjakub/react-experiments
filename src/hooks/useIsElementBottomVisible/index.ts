import {useEffect, useState, RefObject} from 'react';
import isElementBottomInViewPort from 'utils/element-bottom-visible.helpers';

function getCheckIsBottomVisible(
  el: HTMLElement | null,
  setIsVisible: (should: boolean) => void,
) {
  return function () {
    console.log(
      'isVisible checked from handler: ' + isElementBottomInViewPort(el),
    );
    if (isElementBottomInViewPort(el)) {
      setIsVisible(true);
    } else {
      setIsVisible(false);
    }
  };
}

function useIsElementBottomVisible(elementRef: RefObject<HTMLElement>) {
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [isElementChanged, setIsElementChanged] = useState<boolean>(false);
  const element = elementRef.current;

  useEffect(() => {
    const handleWindowChange = getCheckIsBottomVisible(element, isVis => {
      setIsVisible(isVis);
    });
    window.addEventListener('scroll', handleWindowChange);
    window.addEventListener('resize', handleWindowChange);
    return () => {
      window.removeEventListener('scroll', handleWindowChange);
      window.removeEventListener('resize', handleWindowChange);
    };
  }, [element]);

  useEffect(() => {
    if (isElementChanged) {
    
    console.log(
      'isVisible checked from useEffect: ' + isElementBottomInViewPort(element),
    );
      setIsVisible(isElementBottomInViewPort(element));
      setIsElementChanged(false);
    }
  }, [setIsVisible, element, isElementChanged, setIsElementChanged]);

  return {isVisible, setIsElementChanged};
}

export default useIsElementBottomVisible;
