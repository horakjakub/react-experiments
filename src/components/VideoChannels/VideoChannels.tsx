import React, {ReactElement} from 'react';
import VideoChannel from 'components/VideoChannel/VideoChannel';
import Spinner from 'components/Spinner/Spinner';
import useVideoChannels from 'hooks/useVideoChannels/useVideoChannels';
import useDebounce from 'hooks/useDebounce/useDebounce';

export default VideoChannels;

type Props = {
  searchPhrase: string;
};

function VideoChannels({searchPhrase}: Props): ReactElement {
  const {debouncedPhrase} = useDebounce(searchPhrase);
  const {response, isLoading} = useVideoChannels(debouncedPhrase, 5);

  return (
    <>
      {response && !isLoading ? (
        response.map(videoChannelProps => (
          <VideoChannel {...videoChannelProps} key={videoChannelProps.id} />
        ))
      ) : (
        <Spinner />
      )}
    </>
  );
}
