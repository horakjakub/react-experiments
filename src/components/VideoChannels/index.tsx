import React, {ReactElement, useState, RefObject, useEffect} from 'react';
import {differenceWith, isEqual} from 'lodash';
import VideoChannelWithModal from 'components/VideoChannel';
import Spinner from 'components/Spinner';
import useVideoChannels, {mapToVideoChannels} from 'hooks/useVideoChannels';
import {VideoChannelType} from 'common-types/video-channel.type';
import useDebounce from 'hooks/useDebounce';
import useIsElementBottomVisible from 'hooks/useIsElementBottomVisible';
import {VideoChannelsGrid, ResultsPlaceholder} from './styled';

export default VideoChannels;

export function VideoChannelsWrapper({
  children,
  isLoading,
  isNoMoreResults,
  searchPhrase,
}: {
  children: ReactElement;
  isLoading: boolean;
  isNoMoreResults: boolean;
  searchPhrase: string;
}) {
  return (
    <VideoChannelsGrid data-testid="video-channels-wrapper">
      {children}
      {isLoading && (
        <ResultsPlaceholder>
          <Spinner size="6em" color="pink" />
        </ResultsPlaceholder>
      )}
      {isNoMoreResults && (
        <ResultsPlaceholder>
          <p>Oops, there is no results for "{searchPhrase}".</p>
        </ResultsPlaceholder>
      )}
    </VideoChannelsGrid>
  );
}

type Props = {
  searchPhrase: string;
  wrapperRef: RefObject<HTMLElement>;
};

function VideoChannels({searchPhrase, wrapperRef}: Props): ReactElement {
  const [shouldSearch, setShouldSearch] = useState<boolean>(false);
  const [lastSearchPhrase, setLastSearchPhrase] = useState<string | null>(null);
  const [videoChannels, setVideoChannels] = useState<VideoChannelType[]>([]);
  const [nextPageId, setNextPageId] = useState<string | null>(null);
  const {
    isVisible: isBottomBorderVisible,
    setIsElementChanged,
  } = useIsElementBottomVisible(wrapperRef);

  const {debouncedPhrase} = useDebounce(searchPhrase);
  const {response, isLoading} = useVideoChannels(
    shouldSearch && videoChannels.length < 30 ? debouncedPhrase : null,
    20,
    nextPageId,
  );

  useEffect(() => {
    setShouldSearch(
      (isBottomBorderVisible &&
        videoChannels.length &&
        nextPageId !== 'none') ||
        lastSearchPhrase !== debouncedPhrase,
    );

    if (lastSearchPhrase !== debouncedPhrase) {
      setLastSearchPhrase(debouncedPhrase);
      setVideoChannels([]);
    }

    if (response) {
      const {nextPageToken} = response;

      const newChannels = differenceWith(
        mapToVideoChannels(response),
        videoChannels,
        isEqual,
      );

      if (newChannels.length) {
        setVideoChannels([...videoChannels, ...newChannels]);
      }

      if (nextPageToken !== nextPageId) {
        setNextPageId(nextPageToken || 'none');
      }
      setIsElementChanged(true);
    } else {
      setIsElementChanged(false);
    }
  }, [
    response,
    videoChannels,
    nextPageId,
    isBottomBorderVisible,
    debouncedPhrase,
    lastSearchPhrase,
    setIsElementChanged
  ]);

  return (
    <VideoChannelsWrapper
      isLoading={isLoading}
      searchPhrase={searchPhrase}
      isNoMoreResults={nextPageId === 'none'}>
      <>
        {videoChannels.map(videoChannelProps => (
          <VideoChannelWithModal
            {...videoChannelProps}
            key={videoChannelProps.id}
          />
        ))}
      </>
    </VideoChannelsWrapper>
  );
}
