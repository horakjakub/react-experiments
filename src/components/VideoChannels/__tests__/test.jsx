import React, {ChangeEvent} from 'react';
import {
  render,
  fireEvent,
  screen,
  createEvent,
  act,
} from '@testing-library/react';
import useVideoChannels from 'hooks/useVideoChannels';
import useIsElementBottomVisible from 'hooks/useIsElementBottomVisible';
import useDebounce from 'hooks/useDebounce';

import VideoChannels from '../index';

jest.mock('hooks/useVideoChannels');
jest.mock('hooks/useIsElementBottomVisible');
jest.mock('hooks/useDebounce');

describe('<VideoChannels />', () => {
  it('renders correctly', () => {
    const wrapper = document.createElement('div');
    const {container} = render(<VideoChannels wrapperRef={wrapper} />, {
      container: document.body.appendChild(wrapper),
    });

    expect(container).toMatchSnapshot();
  });

  it('should start search when search phrase is passed', () => {
    const wrapper = document.createElement('div');
    useVideoChannels.mockReturnValue({response: {nextPageToken: 'dog'}});

    const {container} = render(
      <VideoChannels wrapperRef={wrapper} searchPhrase="kitty" />,
      {
        container: document.body.appendChild(wrapper),
      },

      //expect().toBe(true);
    );
  });

  it('should clear results previous results when search phrase is changed', () => {});

  it('should search once if a bottom border of <VideoChannels> is not visible', () => {});

  it('should continue search if a bottom border of <VideoChannels /> is visible', () => {});

  it('should stop search if a bottom border of <VideoChannels /> is not visible', () => {});

  it('should start search if a bottom border of <VideoChannels /> container started to be visible', () => {});

  it('should show loading indicator when search is in progress', () => {});

  it('shouldn`t show loading indicator when search is not in progress', () => {});

  it('should show loading indicator when search is in progress', () => {});

  it('shouldn`t show loading indicator when search is not in progress', () => {});
});
