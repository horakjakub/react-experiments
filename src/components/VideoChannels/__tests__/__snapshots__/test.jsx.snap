// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoChannels /> renders correctly 1`] = `
<div>
  <div
    class="sc-fzoLag cQcBgo"
    data-testid="video-channels-wrapper"
  >
    <div
      class="sc-fzoXzr JoTkz"
    >
      <div
        class="sc-AxjAm jcXas"
        color="pink"
      />
    </div>
  </div>
</div>
`;
